### DISCUSSION TITLE
Cannot type anything on Russian

### DISCUSSION THREAD

**Discussion by sindzicat** [olsak/OpTeX#185] [Q&A] [answered]
Hello all!

```tex
\fontfam[Termes]
\typosize[11/13]
\margins/1 a4 (1,1,1,1)cm
\rulang
Привет
\bye
```

output in console:

```
$ optex document.tex
This is LuaTeX, Version 1.18.1 (MiKTeX 24.4 Portable)
 restricted system commands enabled.
(./document.tex This is OpTeX (Olsak's Plain TeX), version <1.15 Jun 2024>
(C:/Apps/miktex-portable/texmfs/install/tex/luatex/optex/base/f-termes.opm
FONT: [Termes] -- \Termes (TeX Gyre Termes fonts based on Times)
 mods:{\caps } vars:{\rm \bf \it \bi } math:{Termes}
MATH-FONT: "[texgyretermes-math]" -- unicode math prepared.

(C:/Apps/miktex-portable/texmfs/install/tex/luatex/optex/base/unimath-codes.opm
(C:/Apps/miktex-portable/texmfs/install/tex/luatex/optex/base/mathclass.opm)
(C:/Apps/miktex-portable/texmfs/install/tex/luatex/optex/base/unimath-table.opm
)))
(C:/Apps/miktex-portable/texmfs/install/tex/generic/hyph-utf8/patterns/tex/hyph-ru.tex)
(C:/Apps/miktex-portable/texmfs/install/tex/luatex/optex/base/lang-data.opm)
[1
Missing character: There is no П (U+041F) in font [texgyretermes-regular]:+tlig;!
{C:/Apps/miktex-portable/texmfs/data/fonts/map/pdftex/pdftex.map}
Missing character: There is no р (U+0440) in font [texgyretermes-regular]:+tlig;!
Missing character: There is no и (U+0438) in font [texgyretermes-regular]:+tlig;!
Missing character: There is no в (U+0432) in font [texgyretermes-regular]:+tlig;!
Missing character: There is no е (U+0435) in font [texgyretermes-regular]:+tlig;!
Missing character: There is no т (U+0442) in font [texgyretermes-regular]:+tlig;!
])<c:/apps/miktex-portable/texmfs/install/fonts/opentype/public/tex-gyre/texgyretermes-regular.otf>
Output written on document.pdf (1 page, 2468 bytes).
Transcript written on document.log.
```

But this font has cyrillic letters:

![image](https://github.com/olsak/OpTeX/assets/33900895/14da96c5-634b-4155-8440-2ad3305ef5e9)

What I do wrong?

Also I'd like to have short code examples in OpTeX user documentation.

**Comment by vlasakm** [comment]
At first, I thought this may require setting the OpenType script correctly to Cyrillic, i.e.

```tex
\_setff{script=cyrl}
\fontfam[Termes]
\typosize[11/13]
\margins/1 a4 (1,1,1,1)cm
\rulang
Привет
\bye
```

But that doesn't seem to help:

```
This is LuaTeX, Version 1.18.0 (mmtex 20240512)
(./test.tex This is OpTeX (Olsak's Plain TeX), version <1.14 Feb 2024>
(/usr/share/mmtex/tex/optex/base/f-termes.opm
FONT: [Termes] -- \Termes (TeX Gyre Termes fonts based on Times)
 mods:{\caps } vars:{\rm \bf \it \bi } math:{Termes}
MATH-FONT: "[texgyretermes-math]" -- unicode math prepared.
(/usr/share/mmtex/tex/optex/base/unimath-codes.opm (/usr/share/mmtex/tex/optex/base/mathclass.opm) (/usr/share/mmtex/tex/optex/base/unimath-table.opm))) (/usr/share/mmtex/tex/hyph-utf8/hyph-ru.tex) (/usr/share/mmtex/tex/optex/base/lang-data.opm) [1
Missing character: There is no П (U+041F) in font [texgyretermes-regular]:+tlig;script=cyrl;!
Missing character: There is no р (U+0440) in font [texgyretermes-regular]:+tlig;script=cyrl;!
Missing character: There is no и (U+0438) in font [texgyretermes-regular]:+tlig;script=cyrl;!
Missing character: There is no в (U+0432) in font [texgyretermes-regular]:+tlig;script=cyrl;!
Missing character: There is no е (U+0435) in font [texgyretermes-regular]:+tlig;script=cyrl;!
Missing character: There is no т (U+0442) in font [texgyretermes-regular]:+tlig;script=cyrl;!
])</usr/share/fonts/tex-gyre/texgyretermes-regular.otf>
Output written on test.pdf (1 page, 2460 bytes).
Transcript written on test.log.
```

I tried to confirm that the font has the characters (e.g. [и, U+0438](https://en.wikipedia.org/wiki/I_(Cyrillic)), but I came out emty handed:

```console
$ otfinfo -u /usr/share/fonts/tex-gyre/texgyretermes-regular.otf | grep 438
uni2260 438 notequal
```

Running `ttx` and checking the output also doesn't seem to show the character. Does the font really have those characters, or am I searching in a wrong way?

If `script=cyrl` font feature is the right path, then search for `\_preplangmore ru` or `\_setff{script=cyrl}` in the [documentation](http://mirrors.ctan.org/macros/optex/doc/optex-doc.pdf) for an idea of how to integrate opentype script switching (or even font switching) into OpTeX language switching.

**Reply by sindzicat** [reply]
Sorry, I don't know TeX so deeply. The only I know this font definitely contains Cyrillic letters.

![image](https://github.com/olsak/OpTeX/assets/33900895/bdf23989-f068-48ac-883f-516e7042e373)

**Reply by Udi-Fogiel** [reply, marked as answer]
I don't think this font contains Cyrillic letters. Word processors don't always use the specified font. If it does not contain certain glyphs they may use a fall back. 
For example
![image](https://github.com/olsak/OpTeX/assets/110533647/c8436d77-0166-4a72-bf12-758271cb7501)
But when I export the file to pdf and inquire the fonts embedded in the file I see
![image](https://github.com/olsak/OpTeX/assets/110533647/85f46fb9-9cab-44da-a7c5-a0682312884c)

**Reply by vlasakm** [reply]
I'll just add that to what @Udi-Fogiel said, that in "TeX" world we usually don't do font fallbacks, so fonts have to explicitly be chosen. If you want to mix latin with Termes and cyrillic with other font, you can do something based on what I had above.

If you have a font in mind it would be possible to try to come up with such a setup for you.

**Reply by Udi-Fogiel** [reply]
@vlasakm see the `fallback` and `multiscript` features described in section 6.1 and font combination described in section 7 of `luaotfload`'s manual.

**Reply by vlasakm** [reply]
@Udi-Fogiel Oh yes, I forgot about that, as I have to admit I have never used it.

**Reply by sindzicat** [reply]
Let's say, this font don't have Cyrillic chars. Is it possible to use Times New Roman in Optex?

**Reply by Udi-Fogiel** [reply]
@sindzicat Yes, it shouldn't be too hard. But before doing a redundant work, did you try New Computer Modern?

**Reply by sindzicat** [reply]
@Udi-Fogiel, I checked [this example](https://mirror.truenetwork.ru/CTAN/fonts/newcomputermodern/doc/newcomputermodern-sample.pdf) and I see that this font looks like Times New Roman. Good advice, thank you. But I'd like to know how to install any font in OpTeX. Is some instruction available on the web? I have no ideas how to find it (what definitely type in Google).
